### DISCUSSION TITLE
pytest-compatible "matcher objects"?

### DISCUSSION THREAD

**Discussion by seanh** [pytest-dev/pytest#12452] [Q&A]
Does anyone know of a library of matcher objects or matcher functions that work nicely with pytest's `assert` statement? Or does anyone have any advice on how to write one?

I'm talking about this sort of thing:

```python
class AnyIterable:
    def __init__(self, containing, only=False, in_order=False, ...):
        ...

    def __eq__(self, other):
        ...

def test_something():
    result = some_function()

    assert result == AnyIterable(containing=[1, 2, 3], only=True)
```

It wouldn't necessarily have to have this particular interface/API design. But the idea is that it provides a way to assert that `result` as an iterable (of any type) containing `1`, `2` and `3`, in any order, but no other values, all in a single easy-to-write assertion.

This particular example is a bit silly. But you can imagine a library with a collection of "matcher objects" like this for matching all sorts of different values (`AnyInstanceOf(SomeModel, with_attrs={"foo": 42, "bar": 23})`, etc), along with providing the user with the ability to write their own matchers.

The aim is to enable writing complex test assertions more fluently.

[PyHamcrest](https://github.com/hamcrest/PyHamcrest) seems to be the leading implementation of this sort of thing in Python but that doesn't look particularly pytest-friendly. It has its own assertion methods like `assert_that(theBiscuit, equal_to(myBiscuit))`, rather than working with normal `assert`'s and comparisons.

Thanks!

**Comment by jwodder** [comment]
I believe [`anys`](https://github.com/jwodder/anys) meets your needs.

Full disclosure: I wrote it.

**Reply by RonnyPfannschmidt** [reply]
@jwodder that looks amazingly close to some ideas we had for matches to integrate in pytest as well

i'd love to collaborate wrt making this a more integral part as the lack of matchers is something that pains since a while

**Reply by RonnyPfannschmidt** [reply]
@jwodder it seems like there may be a need for better assertrepr_compare hooks to better show the differeces for certain matches

**Comment by Suor** [comment]
You can take a look at https://github.com/Suor/pytest-matcher
